nessie/src/main/java/org/apache/iceberg/nessie/TableReference.java

```diff
@@ -64,31 +87,137 @@ public static TableReference parse(TableIdentifier path) {
+
+    private final DateTimeFormatter formatter;
+    private final TemporalQuery<ZonedDateTime> converter;
+
+    FormatOptions(DateTimeFormatter formatter, TemporalQuery<ZonedDateTime> converter) {
+      this.formatter = formatter;
+      this.converter = converter;
+    }
+
+    public ZonedDateTime convert(String timestampStr) {
+      try {
+        return formatter.parse(timestampStr, converter);
+      } catch (DateTimeParseException e) {
+        return null;
+      }
+    }
+  }
+
+  private static Instant parseTimestamp(String timestamp) {
+    ZonedDateTime parsed;
+    for (FormatOptions options : FormatOptions.values()) {
+      parsed = options.convert(modifyTimestamp(timestamp));
+      if (parsed != null) {
+        return endOfPeriod(parsed, timestamp).toInstant();
+      }
+    }
+    throw new IllegalArgumentException(
+        String.format("Cannot parse timestamp: %s is not a legal format. (Use an ISO 8601 compliant string)", timestamp)
+    );
+  }
+
+  private static String modifyTimestamp(String timestamp) {
+    if (timestamp.length() == 7) {
+      // only month. add a day
+      return String.format("%s-01", timestamp);
+    } else if (timestamp.length() == 4) {
+      // only year. add a month and day
+      return String.format("%s-01-01", timestamp);
+    } else {
+      return timestamp;
+    }
+  }
+
+  private static ZonedDateTime endOfPeriod(ZonedDateTime instant, String timestamp) {
+    if (timestamp.length() == 10) {
+      // only date. Move to end of day
+      return endOfPeriod(instant, ChronoUnit.DAYS);
+    } else if (timestamp.length() == 7) {
+      // only month. Move to end of month
+      return endOfPeriod(instant, ChronoUnit.MONTHS);
+    } else if (timestamp.length() == 4) {
+      // only month. Move to end of month
+      return endOfPeriod(instant, ChronoUnit.YEARS);
+    } else {
+      return instant;
+    }
+  }
+
+  private static ZonedDateTime endOfPeriod(ZonedDateTime instant, ChronoUnit unit) {
+    return instant.plus(1, unit).minus(1, ChronoUnit.MICROS);
+  }
+
+  private static ZoneId sparkTimezoneOrUTC() {
+    return sparkTimezone().map(ZoneId::of).orElse(UTC);
+  }
+
+  private static Optional<String> sparkTimezone() {
+    if (sparkAvailable != null) {
+      try {
+        sparkSessionMethod = DynMethods.builder("active")
+            .impl("org.apache.spark.sql.SparkSession").buildStatic();
+        sparkContextMethod = DynMethods.builder("sparkContext")
+            .impl("org.apache.spark.sql.SparkSession").build();
+        sparkConfMethod = DynMethods.builder("conf")
+            .impl("org.apache.spark.SparkContext").build();
+        sparkConfGetMethod = DynMethods.builder("get")
+            .impl("org.apache.spark.SparkConf").build();
+        sparkAvailable = true;
+      } catch (RuntimeException e) {
+        sparkAvailable = false; // spark not on classpath
+      }
+    }
+    if (sparkAvailable != null && sparkAvailable) {
```

Comment:
`sparkAvailable != null` is not redundant because it must be set by L206 or L208. Even if it is not set, `if (sparkAvailable)` should still work

Reply:
agreed. I have split into two `boolean`s now. I don't like relying on the `null`ability of a `Boolean`